File: solutions_5670465267826688_0/java/cjoa/c_dijkstra.java
import java.io.*;

public class c_dijkstra {
   
   public static void main (String[] args) throws IOException {
      (new c_dijkstra()).run();
   }
   
   int[][] MULT_TABLE = new int[8][8];
   public c_dijkstra() {
      MULT_TABLE[0][0] = 0;  MULT_TABLE[0][1] = 1;  MULT_TABLE[0][2] = 2;  MULT_TABLE[0][3] = 3;
      MULT_TABLE[1][0] = 1;  MULT_TABLE[1][1] = 4;  MULT_TABLE[1][2] = 3;  MULT_TABLE[1][3] = 6;
      MULT_TABLE[2][0] = 2;  MULT_TABLE[2][1] = 7;  MULT_TABLE[2][2] = 4;  MULT_TABLE[2][3] = 1;
      MULT_TABLE[3][0] = 3;  MULT_TABLE[3][1] = 2;  MULT_TABLE[3][2] = 5;  MULT_TABLE[3][3] = 4;

      for (int a = 0; a < 4; ++a)
         for (int b = 4; b < 8; ++b)
            MULT_TABLE[a][b] = ( MULT_TABLE[a][b-4] + 4 ) % 8;

      for (int a = 4; a < 8; ++a)
         for (int b = 0; b < 4; ++b)
            MULT_TABLE[a][b] = ( MULT_TABLE[a-4][b] + 4 ) % 8;

      for (int a = 4; a < 8; ++a)
         for (int b = 4; b < 8; ++b)
            MULT_TABLE[a][b] = MULT_TABLE[a-4][b-4];
   }

   static int valueOf(char c) {
      if (c == '1') return 0;
      return c - 'i' + 1;
   }
   
   static String valueOf(int v) {
      String res = "";
      if (v >= 4) {
         res += '-';
         v -= 4;
      }
      if (v == 0) res += '1';
      else res += (char) ('i' + v - 1);
      return res;
   }
   
   void run() throws IOException {
      BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
      int TC = Integer.parseInt(in.readLine());
      for (int tc = 1; tc <= TC; ++tc) {
         String[] tok = in.readLine().split(" ");
         int L = Integer.parseInt(tok[0]);
         int X = Integer.parseInt(tok[1]);
         String S = in.readLine();
         StringBuilder sb = new StringBuilder();
         for (int x = 0; x < X; ++x)
            sb.append(S);
         int cur = 0;
         int stage = 0;
         for (int i = 0; i < sb.length(); ++i) {
            int v = valueOf( sb.charAt(i) );
         //   System.err.print( i + " " + sb.charAt(i) + ": " + valueOf(cur) + " * " + valueOf(v) + " = ");
            cur = MULT_TABLE[cur][v];
         //   System.err.println(valueOf(cur));
            if (cur == 1 && stage == 0)
               stage = 1;
            else if (cur == 3 && stage == 1)
               stage = 2;
         }
         if (cur == 4 && stage == 2) stage = 3;
         String res =  stage == 3 ? "YES" : "NO";
         System.out.printf("Case #%d: %s\n", tc, res);
      }
   }
}

File: solutions_5670465267826688_1/C#/cjoa/c_dijkstra.cs
using System;
using System.Collections.Generic;
using System.Text;

class c_dijkstra {
   static void Main(string[] args) {
      MULT['1','1'] = "1";  MULT['1','i'] = "i";  MULT['1','j'] = "j";  MULT['1','k'] = "k";
      MULT['i','1'] = "i";  MULT['i','i'] = "-1"; MULT['i','j'] = "k";  MULT['i','k'] = "-j";
      MULT['j','1'] = "j";  MULT['j','i'] = "-k"; MULT['j','j'] = "-1"; MULT['j','k'] = "i";
      MULT['k','1'] = "k";  MULT['k','i'] = "j";  MULT['k','j'] = "-i"; MULT['k','k'] = "-1";
      (new c_dijkstra()).Run();
   }

   static string[,] MULT = new string[256,256];

   static string multiply(string s, string t) {
      string prod = MULT[ s[ s.Length-1 ], t[t.Length-1] ];
      if (s[0] == '-' && t[0] == '-' ||
          s[0] != '-' && t[0] != '-')
         return prod;
      return (prod[0] == '-') ? prod.Substring(1) : ("-" + prod);
   }
   
   void Run() {
      int TC = int.Parse(Console.ReadLine());
      for (int tc = 1; tc <= TC; ++tc) {
         string[] tok = Console.ReadLine().Split(' ');
         int L = int.Parse(tok[0]);
         long X = long.Parse(tok[1]);
         string S = Console.ReadLine();
         
         Dictionary< string, long >[] D = new Dictionary< string, long >[3];
         for (int s = 0; s < 3; ++s) {
            D[s] = new Dictionary< string, long >();
            D[s][ "1"] = D[s][ "i"] = D[s][ "j"] = D[s][ "k"] = 
            D[s]["-1"] = D[s]["-i"] = D[s]["-j"] = D[s]["-k"] = -1;
         }

         string cur = "1";
         int stage = 0;
         int x = 0;
         D[stage]["1"] = x;
         while (x < X) {
            for (int i = 0; i < L; ++i) {
               string nxt = multiply(cur, "" + S[i]);
//               Console.Error.WriteLine("{0} {1} {2} : {3} * {4} = {5}",
//                  x, i, x*L + i, cur, "" + S[i], nxt);
               if (nxt == "i" && stage == 0 ||
                   nxt == "k" && stage == 1) ++stage;
               cur = nxt;
            }
            ++x;
            if (D[stage][cur] >= 0)
               break;
            D[stage][cur] = x;
         }
         string res = "NO";
         Console.Error.WriteLine("Stage {0}", stage);
         if (stage == 2) {
            if (x < X) {
               // cycle
               Console.Error.WriteLine(
                  "Repeated after {0} iterations starting from {1}", x, D[stage][cur]);
               long preperiod = D[stage][cur];
               Console.Error.WriteLine("goal = {0}", D[stage]["-1"]);
               long cyclelen = (x - D[stage][cur]);
               if (D[stage]["-1"] >= 0 && (X - preperiod) % cyclelen == D[stage]["-1"] - preperiod)
                  res = "YES";
            }
            else {
               Console.Error.WriteLine("Ended before cycle");
               res = cur == "-1" ? "YES" : "NO";
            }
         }
         Console.WriteLine("Case #{0}: {1}", tc, res);
      }
   }
}